odl/discr/tensor_ops.py

```diff
@@ -0,0 +1,610 @@
+            raise NotImplementedError
+
+        if exponent is None:
+            if self.domain.exponent is None:
+                raise ValueError('cannot determine exponent from {}.'
+                                 ''.format(self.domain))
+            self._exponent = self.domain.exponent
+        elif 0 <= exponent < 1:
+            raise ValueError('exponent between 0 and 1 not allowed.')
+        else:
+            self._exponent = float(exponent)
+
+        # Handle weighting, including sanity checks
+        if weight is None:
+            if hasattr(self.domain.weighting, 'vector'):
+                self._weights = self.domain.weighting.vector
+            elif hasattr(self.domain.weighting, 'const'):
+                self._weights = (self.domain.weighting.const *
+                                 np.ones(len(self.domain)))
+            else:
+                raise ValueError('weighting scheme {!r} of the domain does '
+                                 'not define a weighting vector or constant.'
+                                 ''.format(self.domain.weighting))
+        elif np.isscalar(weight):
+            if weight <= 0:
+                raise ValueError('weighting constant must be positive, got '
+                                 '{}.'.format(weight))
+            self._weights = float(weight) * np.ones(self.domain.size)
+        else:
+            self._weights = np.asarray(weight, dtype='float64')
+            if (not np.all(self.weights > 0) or
+                    not np.all(np.isfinite(self.weights))):
+                raise ValueError('weighting array {} contains invalid '
+                                 'entries.'.format(weight))
+        self._is_weighted = (not np.all(np.array_equiv(self.weights, 1.0)))
+
+    @property
+    def exponent(self):
+        """Exponent ``p`` of this norm."""
+        return self._exponent
+
+    @property
+    def weights(self):
+        """Weighting vector of this norm."""
+        return self._weights
+
+    @property
+    def is_weighted(self):
+        """`True` if weighting is not 1 or all ones."""
+        return self._is_weighted
+
+    def _call(self, f, out):
+        """Implement ``self(f, out)``."""
+        if len(self.domain.shape) > 1:
+            raise NotImplementedError
+
+        if self.exponent == 1.0:
+            self._call_vecfield_1(f, out)
+        elif self.exponent == float('inf'):
+            self._call_vecfield_inf(f, out)
+        else:
+            self._call_vecfield_p(f, out)
+
+    def _call_vecfield_1(self, vf, out):
+        """Implement ``self(vf, out)`` for exponent 1."""
+        vf[0].ufunc.absolute(out=out)
+        if self.is_weighted:
+            out *= self.weights[0]
+
+        if self.domain.size == 1:
+            return
+
+        tmp = self.range.element()
+        for fi, w in zip(vf[1:], self.weights[1:]):
+            fi.ufunc.absolute(out=tmp)
+            if self.is_weighted:
+                tmp *= w
```

Comment:
How would you do that? The `w` is different for each product space component, so I guess we have to do one weighting component at a time.